srml/balances/src/lib.rs

```diff
@@ -311,9 +320,14 @@ impl<T: Trait> Module<T> {
 			None => return Err("destination balance too high to receive value"),
 		};
 
+		if !can_kill {
+			ensure!(new_from_balance >= Self::existential_deposit(), "balance too low to send value")
+		}
+
 		if transactor != dest {
-			Self::set_free_balance(transactor, new_from_balance);
-			Self::decrease_total_stake_by(fee);
+			T::ChargeFee::charge_fee(transactor, fee)?;
+
+			Self::set_free_balance(transactor, new_from_balance - fee);
```

Comment:
Why is the fee subtracted again? What protects us from overflows?

Reply:
@tomusdrw If not subtracted, `fee` will actually be charged twice, since `charge_fee` has already removed fee from balance. Because `fee` was added to `new_from_balance` (by calculating `liability`), the subtraction here cannot fail.

Reply:
If I'm not mistaken it's the other way around:
```rust
let liability = match value.checked_add(&fee)...;
let new_from_balance = match from_balance.checked_sub(&liability)...;
T::ChargeFee::charge_fee(transactor, fee)?;
Self::set_free_balance(transactor, new_from_balance - fee);
```
The fee is *subtracted* from `from_balance` and then again subtracted from `free_balance`. So unless I'm missing something: `new_from_balance = from_balance - value - fee - fee`

Reply:
@tomusdrw You're right, my mistake. Will update.

Reply:
Updated. Also added two unit tests for balance transfer, to cover fee charge.